### PR TITLE
Tm 180 export application details csv

### PR DESCRIPTION
**Pull request by sscaife**

### CODE REVIEW

File: src/SFA.DAS.LevyTransferMatching.UnitTests/Application/Queries/Pledges/GetApplicationsForDownload/GetApplicationsForDownloadQueryHandlerTests.cs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+namespace SFA.DAS.LevyTransferMatching.UnitTests.Application.Queries.Pledges.GetApplicationsForDownload
+{
+    public class GetApplicationsForDownloadQueryHandlerTests
+    {
+        private GetApplicationsForDownloadQueryHandler _handler;
+        private Mock<ILevyTransferMatchingService> _service;
+        private Mock<ICoursesApiClient<CoursesApiConfiguration>> _coursesApiClient;
+        private Mock<IReferenceDataService> _referenceDataService;
+        //private GetApplicationsForDownloadQuery _query;
```
Comment (chrisfoster76) [inline src/SFA.DAS.LevyTransferMatching.UnitTests/Application/Queries/Pledges/GetApplicationsForDownload/GetApplicationsForDownloadQueryHandlerTests.cs:31]: Could we remove the commented lines?

File: src/SFA.DAS.LevyTransferMatching/Application/Queries/Pledges/GetApplicationsForDownload/GetApplicationsForDownloadQueryHandler.cs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+            if (!applications.Applications.Any())
+            {
+                return new GetApplicationsForDownloadQueryResult
+                {
+                    Applications = new List<ApplicationForDownloadModel>()
+                };
+            }
+
+            var standardTasks = RetrieveStandardTasks(applications);
```
Comment (chrisfoster76) [inline src/SFA.DAS.LevyTransferMatching/Application/Queries/Pledges/GetApplicationsForDownload/GetApplicationsForDownloadQueryHandler.cs:52]: Standard info can be retrieved from the application directly now (from inner api)

File: src/SFA.DAS.SharedOuterApi/Models/Application.cs [hunk 3]
```diff
@@ -24,12 +24,16 @@ public class Application
         public Standard Standard { get; set; }
         public bool IsNamePublic { get; set; }
         public string Status { get; set; }
+        public IEnumerable<string> PledgeLocations { get; set; }
```
Comment (chrisfoster76) [inline src/SFA.DAS.SharedOuterApi/Models/Application.cs:27]: It feels as though too much is being added to the application object coming back from the inner api - perhaps a separate call to get the pledge would be better?

**Review by chrisfoster76 — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.LevyTransferMatching.UnitTests/Application/Queries/Pledges/GetApplicationsForDownload/GetApplicationsForDownloadQueryHandlerTests.cs (hunk 1 above)
Comment (sscaife) [reply src/SFA.DAS.LevyTransferMatching.UnitTests/Application/Queries/Pledges/GetApplicationsForDownload/GetApplicationsForDownloadQueryHandlerTests.cs:31]: yeah sure

File: src/SFA.DAS.LevyTransferMatching/Application/Queries/Pledges/GetApplications/GetApplicationsQueryHandler.cs [hunk 4]
```diff
@@ -1,10 +1,14 @@
 ﻿using System.Collections.Generic;
 using System.Linq;
+using System.Security.Cryptography.X509Certificates;
```
Comment (chrisfoster76) [inline src/SFA.DAS.LevyTransferMatching/Application/Queries/Pledges/GetApplications/GetApplicationsQueryHandler.cs:3]: Presumably this is an accident

File: src/SFA.DAS.LevyTransferMatching/Application/Queries/Pledges/GetApplications/GetApplicationsQueryHandler.cs [hunk 5]
```diff
@@ -37,12 +41,8 @@ public async Task<GetApplicationsQueryResult> Handle(GetApplicationsQuery reques
...
 
                 var applicationResult = (GetApplicationsQueryResult.Application)application;
-
-                applicationResult.IsLocationMatch = !pledgeResponse.Locations.Any() || application.Locations.Any();
-                applicationResult.IsSectorMatch = !pledgeResponse.Sectors.Any() || application.Sectors.Any(x => pledgeResponse.Sectors.Contains(x));
-                applicationResult.IsJobRoleMatch = !pledgeResponse.JobRoles.Any() || roles.Any(r => r.Description == application.StandardRoute);
-                applicationResult.IsLevelMatch = !pledgeResponse.Levels.Any() || pledgeResponse.Levels.Select(x => char.GetNumericValue(x.Last())).Contains(application.StandardLevel);
-
+                
+                AddPledgeDetailsToApplication(applicationResult, pledgeResponse, application, roles);
```
Comment (chrisfoster76) [inline src/SFA.DAS.LevyTransferMatching/Application/Queries/Pledges/GetApplications/GetApplicationsQueryHandler.cs:45]: As much as I like the implicit operators, I think this stitching together of these bits is a bit messy (i.e. using the implicit operator to do "most of" the mapping and then doing a bit more elsewhere), though I appreciate the "match" fields were being tacked on like that previously.

Could we just remove the implicit operator here and add a mapping method which takes all the required params so the application result gets instantiated and all properties set in one go?

File: src/SFA.DAS.LevyTransferMatching.Api/Models/Pledges/GetApplicationsResponse.cs [hunk 6]
```diff
@@ -1,5 +1,8 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
```
Comment (chrisfoster76) [inline src/SFA.DAS.LevyTransferMatching.Api/Models/Pledges/GetApplicationsResponse.cs:4]: Guessing this is unused

**Review by chrisfoster76 — CHANGES_REQUESTED** [review]
Just one comment about some mapping that could be tidied up, and a couple of unused usings.

File: src/SFA.DAS.LevyTransferMatching/Application/Queries/Pledges/GetApplicationsForDownload/GetApplicationsForDownloadQueryHandler.cs (hunk 2 above)
Comment (sscaife) [reply src/SFA.DAS.LevyTransferMatching/Application/Queries/Pledges/GetApplicationsForDownload/GetApplicationsForDownloadQueryHandler.cs]: removed the file now

File: src/SFA.DAS.LevyTransferMatching/Application/Queries/Pledges/GetApplications/GetApplicationsQueryHandler.cs (hunk 5 above)
Comment (sscaife) [reply src/SFA.DAS.LevyTransferMatching/Application/Queries/Pledges/GetApplications/GetApplicationsQueryHandler.cs]: added a factory method that takes the parameters and builds the object

**Review by ConorTill — APPROVED** [review]

**Review by chrisfoster76 — APPROVED** [review]

**Pull request merged**